src/util/ByteOrder.hxx

```diff
@@ -263,7 +264,7 @@ public:
 			x = ByteSwap16(x);
 		return x;
 	}
-};
+} __attribute__((packed));
```

Comment:
This is a proprietary compiler extension that should not be used unconditionally, but why use it here??

Reply:
Also, adding it to this existing structure is unrelated to this commit and is not even explained in the commit message.
Please don't hide unrelated changes. I'll find crap like that!

Reply:
When I used the PackedLE32 (or others) in my structures (which are using this attribute to ensure no padding), the compiler gave error that the attribute (of the struct) is ignored because it contains "unpacked non-POD". Adding the attribute to the classes fixed that.

Reply:
So you have no idea what you're doing, and you did some random thing and this made the compiler shut up, so you figured this must be the solution to your problem?

Reply:
The __attribute__((packed)) is used in an example of data structure in the protocol documentation from LX Navigation. So it is not entirely random. But you are right that I have no deep understanding of what this does, just that it somehow ensures that there is no padding between the variables.

Reply:
I previously used this for my structures in the LXEosDevic, which was already reviewed by you I think.

Reply:
Should I rather use gcc_packed (as in the LXNAV driver)? What is the difference?

Reply:
The LXNAV drive is using the unit and int types in the structures, not the PackedBE32 and others. I think they are actually used nowhere in the code.

Reply:
> The __attribute__((packed)) is used in an example of data structure in the protocol documentation from LX Navigation.

And LX Navigation's documentation follows which standard?

> So it is not entirely random.

You're missing the point, and it sounds like you don't know what I'm talking about. You havn't even tried to answer my question. I asked why you use it. You have not explained.

Saying it's in the LX Navigation documentation in example code doesn't answer the question, unless you want to tell me that you just copy somebody else's code without understanding it. Code that nobody can explain will never be merged into XCSoar.

> Should I rather use gcc_packed (as in the LXNAV driver)?

Why would you want to do that? This simple question, you're still sidestepping it wherever you can, instead of answering it.

Reply:
My understanding is: When I have a structure that contains variables with sizes different than 4 bytes (on 32-bit platform), the compiler may decide to pad the less-than-4-byte variable with empty bytes so that the next variable begins at an address that is multiple of 4, because the CPU can work with that efficiently. This is unwanted for my data structures, because it must exactly match the communication protocol bytes. I already wrote 

> it somehow ensures that there is no padding between the variables

I know about two ways to achieve this: gcc_packed and the attribute((packed)) 
I don't know which of them is preferred. 

And it seems that the attribute cannot work on a struct that contains "unpacked non-POD", so I added the attribute to the PackedBE32 and others too. It makes no practical difference for them (if they had any padding, the static_assert of their size would fail), but it 

> made the compiler shut up

Reply:
BTW the reason I chose atribute((packed)) rather than gcc_packed is that when using gcc_packed, my IDE was complaining about the asserts checking the size of the structs. Of course, this is just some wrong setting of the IDE, which is not a good reason for deciding which to use. But I wasn't aware of any reason to prefer one over the other anyway.

Reply:
Your understanding is not correct.

Reply:
You can help me better understand it by telling me if the following claims are true or not:

The compiler can place unused bytes between variables to ensure that variables are properly aligned in the memory.

Aligned means that they can be accessed efficiently. Misaligned causes lower performance or even exceptions.

On a 32-bit platform, properly aligned variables will start at a memory address that is multiple of 4

The data structure used for the communication protocol must not contain any padding.

This can be achieved by telling the compiler not to do any alignment in the structure.

Which is the purpose of the "packed" attribute.

Structure with this attribute cannot have "unpacked non-POD" members.

These are classes or structures that do not have the packed attribute.

So I need to add the attribute to the Packed32BE (and other types I use)

Reply:
No, you can't just put the burden on me and throw this wall of statements at me, making it my responsibility to verify them. I don't have time for that.
And no again, PackedBE32 doesn't need the "padded" attribute.
I'm tired of this PR, I'm going to unsubscribe. When you're completely done, have fixed all the problems and it's no longer in "Draft" status, you may ask for another review, but until then, I decide my time is too precious for you to waste it.

Reply:
Honestly, I was not expecting you to review it while it is still in "draft" status...

Reply:
I still don't understand what is going on

The Packed32BE indeed is 4 byte even if the "packed" attribute is not used. 

However, when I try to use the Packed32BE in the struct of data that are transmitted to vario (this struct obviously must be packed), the compiler will say that the packed attribute is ignored because the structure contains "unpacked non-POD". 

Adding the packed attribute to Packed32BE solves that, but you say that it is not correct (or at least not needed) to do so. So what would be a good solution? The Packed32BE is not used anywhere else in the code; I thought this was the intended use-case of Packed32BE, was it? Was it ever used (and working)?

Reply:
I still think that the attribute needs to be there (to be able to use the Packed32BE inside another struct that is packed)

Reply:
So if I try to answer my previous questions:

> The compiler can place unused bytes between variables to ensure that variables are properly aligned in the memory.

True (or at least close-enough to true)

> Aligned means that they can be accessed efficiently. Misaligned causes lower performance or even exceptions.

True (or at least close-enough to true)

> On a 32-bit platform, properly aligned variables will start at a memory address that is multiple of 4

False, it is a multiple of the size of the data type, so for example `uint8_t a,b,c,d;` will have no padding.

> The data structure used for the communication protocol must not contain any padding.

True

> This can be achieved by telling the compiler not to do any alignment in the structure.
> Which is the purpose of the "packed" attribute.

True (or at least close-enough to true)

> Structure with this attribute cannot have "unpacked non-POD" members.

That's what the compiler is complaining about

> These are classes or structures that do not have the packed attribute.

Maybe false, but seems like the compiler will be satisfied only if I add the attribute

> So I need to add the attribute to the Packed32BE (and other types I use)

That's what I think. If it is not true, I would be glad for any suggestions on how to solve it in other way.